crypto/fipsmodule/rand/entropy/tree_drbg_jitter_entropy.c

```diff
@@ -0,0 +1,465 @@
+  if (CRYPTO_get_ube_generation_number(&current_generation_number) != 1 &&
+      current_generation_number != tree_jitter_drbg->generation_number) {
+    tree_jitter_drbg->generation_number = current_generation_number;
+    return 1;
+  }
+
+  if (tree_jitter_drbg->generate_calls_since_seed + 1 > tree_jitter_drbg->max_generate_calls) {
+    return 1;
+  }
+
+  return 0;
+}
+
+// tree_jitter_drbg_derive_seed generates a CTR_DRBG_ENTROPY_LEN byte seed from
+// the DRBG configured in |tree_jitter_drbg|. The generated bytes are returned
+// in |seed_out|.
+//
+// |tree_jitter_drbg_derive_seed| automatically handles reseeding the
+// associated DRBG if required. In addition, if UBE detection is not supported
+// prediction resistance is used to ensure bytes are generated safely.
+static void tree_jitter_drbg_derive_seed(
+  struct tree_jitter_drbg_t *tree_jitter_drbg,
+  uint8_t seed_out[CTR_DRBG_ENTROPY_LEN]) {
+
+  if (tree_jitter_drbg == NULL) {
+    abort();
+  }
+
+  if (tree_jitter_check_drbg_must_reseed(tree_jitter_drbg) == 1) {
+    uint8_t seed_drbg[CTR_DRBG_ENTROPY_LEN];
+    if (tree_jitter_drbg->is_global == 1) {
+      tree_jitter_get_root_seed(tree_jitter_drbg, seed_drbg);
+    } else {
+      CRYPTO_STATIC_MUTEX_lock_write(global_seed_drbg_lock_bss_get());
+      tree_jitter_drbg_derive_seed(*global_seed_drbg_bss_get(), seed_drbg);
+      CRYPTO_STATIC_MUTEX_unlock_write(global_seed_drbg_lock_bss_get());
+    }
+
+    if (CTR_DRBG_reseed(&(tree_jitter_drbg->drbg), seed_drbg, NULL, 0) != 1) {
+      abort();
+    }
+    OPENSSL_cleanse(seed_drbg, CTR_DRBG_ENTROPY_LEN);
+    tree_jitter_drbg->reseed_calls_since_initialization += 1;
+    tree_jitter_drbg->generate_calls_since_seed = 0;
+  }
+
+  uint8_t pred_resistance[RAND_PRED_RESISTANCE_LEN];
+  size_t pred_resistance_len = 0;
+  tree_jitter_drbg_maybe_get_pred_resistance(tree_jitter_drbg,
+    pred_resistance, &pred_resistance_len);
+
+  OPENSSL_STATIC_ASSERT(CTR_DRBG_ENTROPY_LEN <= CTR_DRBG_MAX_GENERATE_LENGTH,
+    CTR_DRBG_ENTROPY_LEN_is_too_large_compared_to_CTR_DRBG_MAX_GENERATE_LENGTH)
+
+  if (!CTR_DRBG_generate(&(tree_jitter_drbg->drbg), seed_out, CTR_DRBG_ENTROPY_LEN,
+        pred_resistance, pred_resistance_len)) {
+    abort();
+  }
+  OPENSSL_cleanse(pred_resistance, RAND_PRED_RESISTANCE_LEN);
+  tree_jitter_drbg->generate_calls_since_seed += 1;
+}
+
+// tree_jitter_drbg_derive_seed generates a CTR_DRBG_ENTROPY_LEN byte seed from
+// the DRBG configured in |entropy_source|. The generated bytes are returned
+// in |seed_out|. This function is the entry point for generating output from
+// the tree DRBG.
+//
+// Return 1 on success and 0 otherwise.
+int tree_jitter_get_seed(const struct entropy_source_t *entropy_source,
+  uint8_t seed_out[CTR_DRBG_ENTROPY_LEN]) {
+
+  GUARD_PTR_ABORT(entropy_source);
+  GUARD_PTR_ABORT(seed_out);
+
+  struct tree_jitter_drbg_t *tree_jitter_drbg_thread =
+    (struct tree_jitter_drbg_t *) entropy_source->state;
```

Comment:
Because each implementation of interface defined by `entropy_source_method` might need a state specific to the implementation. It's pretty standard to do it this way IMO. The other option is to start implementing type definitions.